### PR TITLE
Structured text language support

### PR DESCRIPTION
**Pull request by Serhioromano**
As discussed in #1923 here is PR

### CODE REVIEW

File: components/prism-iecst.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+			lookbehind: true
+		},
+		{
+			pattern: /(^|[^\\])\(\*[\s\S]*?(?:\*\)|$)/,
+			lookbehind: true
+		},
+		{
+			pattern: /(^|[^\\])\{[\s\S]*?(?:\}|$)/,
+			lookbehind: true
+		},
```
Comment (RunDevelopment) [inline components/prism-iecst.js:14]: It's more efficient to combine these patterns.

```suggestion
		{
			pattern: /(^|[^\\])(?:\/\*[\s\S]*?(?:\*\/|$)|\(\*[\s\S]*?(?:\*\)|$)|\{[\s\S]*?(?:\}|$))/,
			lookbehind: true
		},
```

File: components/prism-iecst.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+			pattern: /(^|[^\\:])\/\/.*/,
+			lookbehind: true,
+			greedy: true
+		}
+    ],
+    'boolean': /\b(?:true|false|null)\b/,
+	'class-name': /\b(?:END_)?(PROGRAM|CONFIGURATION|INTERFACE|FUNCTION_BLOCK|FUNCTION|ACTION|TRANSITION|TYPE|STRUCT|(?:INITIAL_)?STEP|NAMESPACE|LIBRARY|CHANNEL|FOLDER|RESOURCE|VAR_(?:GLOBAL|INPUT|PUTPUT|IN_OUT|ACCESS|TEMP|EXTERNAL|CONFIG)|VAR|METHOD|PROPERTY)\b/i,
+	'keyword': /\b(?:(?:END_)?(?:IF|WHILE|REPEAT|CASE|FOR)|ELSE|FROM|THEN|ELSIF|DO|TO|BY|PRIVATE|PUBLIC|PROTECTED|CONSTANT|RETURN|EXIT|CONTINUE|GOTO|JMP|AT|RETAIN|NON_RETAIN|TASK|WITH|UNTIL|USING|EXTENDS|IMPLEMENTS|GET|SET|__TRY|__CATCH|__FINALLY|__ENDTRY)\b/,
+	
+    'string': {
```
Comment (RunDevelopment) [inline components/prism-iecst.js:25]: Please move this after `comment`.

The placement of `string` doesn't matter from a correctness point of view but it's more performant to have all greedy tokens at the start of the language definition.

**Comment by Serhioromano** [comment]
I've solved conflicts

File: components/prism-iecst.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		pattern: /(["'])(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/,
+		greedy: true
+    },
+    'variable': /\b(?:AT|BOOL|BYTE|(?:D|L)?WORD|U?(?:S|D|L)?INT|L?REAL|TIME(?:_OF_DAY)?|TOD|DT|DATE(?:_AND_TIME)?|STRING|ARRAY|ANY|POINTER)\b/,
+    'symbol': /%(?:I|Q|M)(?:X|B|W|D|L)[0-9\.]*|%(?:I|Q)[0-9\.]*/,
+    'number': /\b(?:16#[0-9ABCDEF]*)|\b(?:2#[01_]*)|\b(?:(?:T|D|DT|TOD)#[0-9_mshmd:]*)|\b(?:[A-Z]*\#[0-9.,_]*)|\b0x[\da-f]+\b|(?:\b\d+\.?\d*|\B\.\d+)(?:e[+-]?\d+)?/i,
+    'function': /\b\w+(?=\()/,
```
Comment (RunDevelopment) [inline components/prism-iecst.js:32]: The `\b` is unnecessary.

`\b\w+(?=\()` is equivalent to `(?<!\w)\w+(?=\()` in this context. Because of the greedy left-to-right matching, we know that the character (if any) before the `\w+` doesn't match `\w`, so the `(?<!\w)` (caused by the `\b`) is unnecessary.

File: components/prism-iecst.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	'class-name': /\b(?:END_)?(PROGRAM|CONFIGURATION|INTERFACE|FUNCTION_BLOCK|FUNCTION|ACTION|TRANSITION|TYPE|STRUCT|(?:INITIAL_)?STEP|NAMESPACE|LIBRARY|CHANNEL|FOLDER|RESOURCE|VAR_(?:GLOBAL|INPUT|PUTPUT|IN_OUT|ACCESS|TEMP|EXTERNAL|CONFIG)|VAR|METHOD|PROPERTY)\b/i,
+	'keyword': /\b(?:(?:END_)?(?:IF|WHILE|REPEAT|CASE|FOR)|ELSE|FROM|THEN|ELSIF|DO|TO|BY|PRIVATE|PUBLIC|PROTECTED|CONSTANT|RETURN|EXIT|CONTINUE|GOTO|JMP|AT|RETAIN|NON_RETAIN|TASK|WITH|UNTIL|USING|EXTENDS|IMPLEMENTS|GET|SET|__TRY|__CATCH|__FINALLY|__ENDTRY)\b/,
+	
+    'string': {
+		pattern: /(["'])(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/,
+		greedy: true
+    },
+    'variable': /\b(?:AT|BOOL|BYTE|(?:D|L)?WORD|U?(?:S|D|L)?INT|L?REAL|TIME(?:_OF_DAY)?|TOD|DT|DATE(?:_AND_TIME)?|STRING|ARRAY|ANY|POINTER)\b/,
+    'symbol': /%(?:I|Q|M)(?:X|B|W|D|L)[0-9\.]*|%(?:I|Q)[0-9\.]*/,
+    'number': /\b(?:16#[0-9ABCDEF]*)|\b(?:2#[01_]*)|\b(?:(?:T|D|DT|TOD)#[0-9_mshmd:]*)|\b(?:[A-Z]*\#[0-9.,_]*)|\b0x[\da-f]+\b|(?:\b\d+\.?\d*|\B\.\d+)(?:e[+-]?\d+)?/i,
```
Comment (RunDevelopment) [inline components/prism-iecst.js:31]: 1. `[0-9ABCDEF]` -> `[\da-f]`
2. `[0-9_mshmd:]` -> `[0-9_shmd:]`
3. I combined `\b(?:16#[0-9ABCDEF]*)`, `\b(?:2#[01_]*)`, and `\b0x[\da-f]+\b` into `\b(?:16#[\da-f]+|2#[01_]+|0x[\da-f]+)\b`.

```suggestion
    'number': /\b(?:16#[\da-f]+|2#[01_]+|0x[\da-f]+)\b|\b(?:T|D|DT|TOD)#[\d_shmd:]*|\b[A-Z]*\#[\d.,_]*|(?:\b\d+\.?\d*|\B\.\d+)(?:e[+-]?\d+)?/i,
```

File: components/prism-iecst.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    'boolean': /\b(?:true|false|null)\b/,
+	'class-name': /\b(?:END_)?(PROGRAM|CONFIGURATION|INTERFACE|FUNCTION_BLOCK|FUNCTION|ACTION|TRANSITION|TYPE|STRUCT|(?:INITIAL_)?STEP|NAMESPACE|LIBRARY|CHANNEL|FOLDER|RESOURCE|VAR_(?:GLOBAL|INPUT|PUTPUT|IN_OUT|ACCESS|TEMP|EXTERNAL|CONFIG)|VAR|METHOD|PROPERTY)\b/i,
+	'keyword': /\b(?:(?:END_)?(?:IF|WHILE|REPEAT|CASE|FOR)|ELSE|FROM|THEN|ELSIF|DO|TO|BY|PRIVATE|PUBLIC|PROTECTED|CONSTANT|RETURN|EXIT|CONTINUE|GOTO|JMP|AT|RETAIN|NON_RETAIN|TASK|WITH|UNTIL|USING|EXTENDS|IMPLEMENTS|GET|SET|__TRY|__CATCH|__FINALLY|__ENDTRY)\b/,
+	
+    'string': {
+		pattern: /(["'])(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/,
+		greedy: true
+    },
+    'variable': /\b(?:AT|BOOL|BYTE|(?:D|L)?WORD|U?(?:S|D|L)?INT|L?REAL|TIME(?:_OF_DAY)?|TOD|DT|DATE(?:_AND_TIME)?|STRING|ARRAY|ANY|POINTER)\b/,
+    'symbol': /%(?:I|Q|M)(?:X|B|W|D|L)[0-9\.]*|%(?:I|Q)[0-9\.]*/,
```
Comment (RunDevelopment) [inline components/prism-iecst.js:30]:
```suggestion
    'symbol': /%[IQM][XBWDL][\d.]*|%[IQ][\d.]*/,
```

File: components/prism-iecst.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    'number': /\b(?:16#[0-9ABCDEF]*)|\b(?:2#[01_]*)|\b(?:(?:T|D|DT|TOD)#[0-9_mshmd:]*)|\b(?:[A-Z]*\#[0-9.,_]*)|\b0x[\da-f]+\b|(?:\b\d+\.?\d*|\B\.\d+)(?:e[+-]?\d+)?/i,
+    'function': /\b\w+(?=\()/,
+    'operator': /\b(:?S?R?=>?|\+|\^|\-|&&?|\*\*?|\/|<=?|>=?|OR|AND|MOD|NOT|XOR|LE|GE|EQ|NE|GE|LT)\b/,
+    'punctuation': /[\(\);]/,
```
Comment (RunDevelopment) [inline components/prism-iecst.js:34]:
```suggestion
    'punctuation': /[();]/,
```

File: components/prism-iecst.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	
+    'string': {
+		pattern: /(["'])(?:\\(?:\r\n|[\s\S])|(?!\1)[^\\\r\n])*\1/,
+		greedy: true
+    },
+    'variable': /\b(?:AT|BOOL|BYTE|(?:D|L)?WORD|U?(?:S|D|L)?INT|L?REAL|TIME(?:_OF_DAY)?|TOD|DT|DATE(?:_AND_TIME)?|STRING|ARRAY|ANY|POINTER)\b/,
+    'symbol': /%(?:I|Q|M)(?:X|B|W|D|L)[0-9\.]*|%(?:I|Q)[0-9\.]*/,
+    'number': /\b(?:16#[0-9ABCDEF]*)|\b(?:2#[01_]*)|\b(?:(?:T|D|DT|TOD)#[0-9_mshmd:]*)|\b(?:[A-Z]*\#[0-9.,_]*)|\b0x[\da-f]+\b|(?:\b\d+\.?\d*|\B\.\d+)(?:e[+-]?\d+)?/i,
+    'function': /\b\w+(?=\()/,
+    'operator': /\b(:?S?R?=>?|\+|\^|\-|&&?|\*\*?|\/|<=?|>=?|OR|AND|MOD|NOT|XOR|LE|GE|EQ|NE|GE|LT)\b/,
```
Comment (RunDevelopment) [inline components/prism-iecst.js:33]: Could you please add a test for this pattern as well?

File: components/prism-iecst.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    'punctuation': /[\(\);]/,
+    'type': {
+		pattern: /#/i,
```
Comment (RunDevelopment) [inline components/prism-iecst.js:36]:
```suggestion
		pattern: /#/,
```

**Comment by Serhioromano** [comment]
I do not know why checks fail. It does work for local when I run npm run test****

**Comment by RunDevelopment** [comment]
What is `npm run test****`?
If you run `npm test` you should see [this](https://travis-ci.org/github/PrismJS/prism/jobs/674079095#L5251):

```
  2 failing
  1) Patterns of 'iecst'
       - should have a capturing group if lookbehind is set to true:
     Error: Token <languages>.iecst.type.pattern: The pattern is set to 'lookbehind: true' but does not have a capturing group.
      at forEachPattern (tests/pattern-tests.js:124:10)
      at Context.<anonymous> (tests/pattern-tests.js:184:3)
  2) Patterns of 'iecst'
       - should not have unused capturing groups:
     Error: Token <languages>.iecst["class-name"]: Unused capturing group (PROGRAM|CONFIGURATION|INTERFACE|FUNCTION_BLOCK|FUNCTION|ACTION|TRANSITION|TYPE|STRUCT|(?:INITIAL_)?STEP|NAMESPACE|LIBRARY|CHANNEL|FOLDER|RESOURCE|VAR_(?:GLOBAL|INPUT|PUTPUT|IN_OUT|ACCESS|TEMP|EXTERNAL|CONFIG)|VAR|METHOD|PROPERTY). All capturing groups have to be either referenced or used as a Prism lookbehind group.
      at forEachPattern (tests/pattern-tests.js:124:10)
      at Context.<anonymous> (tests/pattern-tests.js:261:3)
```

**Comment by Serhioromano** [comment]
This is how I run tests

Snag_ea3efa

**Comment by Serhioromano** [comment]
By reading your messages I applied some changes. I didn't know where to see those messages.

File: plugins/show-language/prism-show-language.js [hunk 9]
```diff
@@ -1,12 +1,10 @@
 (function () {
-
-	if (typeof self === 'undefined' || !self.Prism || !self.document) {
+	if (typeof self === "undefined" || !self.Prism || !self.document) {
 		return;
 	}
 
 	if (!Prism.plugins.toolbar) {
-		console.warn('Show Languages plugin loaded before Toolbar plugin.');
-
```
Comment (RunDevelopment) [inline plugins/show-language/prism-show-language.js:9]: Please revert these changes. While the file might need a cleanup, it shouldn't be part of this PR.

The only change to this file should be this:

```diff
 		"ichigojam": "IchigoJam",
+		"iecst": "Structured Text (IEC 61131-3)",
 		"inform7": "Inform 7",
```

Comment (Serhioromano) [reply plugins/show-language/prism-show-language.js:9]: How? when I run npx gulp it rebuild this file and make them all one line.

Comment (Serhioromano) [reply plugins/show-language/prism-show-language.js:9]: And yet, I am not sure how to revert a single file change. Out of my knowledge. I'll try to google it.

**Comment by RunDevelopment** [comment]
> This is how I run tests

This only checks your language against your `.test` files. We also have a few other tests which scan all patterns for potential errors (like the unnecessary `lookbehind: true`).
`npm test` runs all tests for all languages and basically verifies that Prism is fully functional.

> By reading your messages I applied some changes. I didn't know where to see those messages.

What I showed you was the build log of Travis CI. To see this, click on the red X behind a commit on GitHub, click Details, choose any red (failed) build job, scroll waaaaay down.

File: plugins/show-language/prism-show-language.js (hunk 9 above)
Comment (RunDevelopment) [reply plugins/show-language/prism-show-language.js:9]: The easiest way would probably to copy-paste the content of the current [`prism-show-language.js`](https://github.com/PrismJS/prism/blob/master/plugins/show-language/prism-show-language.js) in master and run `npx gulp`.

There's probably also some fancy git-way of doing it, but I'm not a git expert...

**Comment by Serhioromano** [comment]
I'll create a new PR. This copy of prism was forked years ago and propbably rebase did not work well.

**Comment by RunDevelopment** [comment]
New PR: #2311
